ami/trainers/i_jepa_trainer.py

```diff
@@ -0,0 +1,171 @@
+
+from ami.data.buffers.buffer_names import BufferNames
+from ami.data.buffers.random_data_buffer import RandomDataBuffer
+from ami.data.interfaces import ThreadSafeDataUser
+from ami.models.i_jepa import (
+    IJEPAEncoder,
+    IJEPAPredictor,
+    repeat_patches_along_with_batch_axis,
+    select_patches_by_indices,
+)
+from ami.models.model_names import ModelNames
+from ami.models.model_wrapper import ModelWrapper
+from ami.tensorboard_loggers import StepIntervalLogger
+
+from .base_trainer import BaseTrainer
+from .components.i_jepa_mask_collator import IJEPAMultiBlockMaskCollator
+
+
+class IJEPATrainer(BaseTrainer):
+    def __init__(
+        self,
+        partial_dataloader: partial[DataLoader[torch.Tensor]],
+        partial_optimizer: partial[Optimizer],
+        i_jepa_mask_collator: IJEPAMultiBlockMaskCollator,
+        device: torch.device,
+        logger: StepIntervalLogger,
+        max_epochs: int = 1,
+        minimum_dataset_size: int = 1,
+        minimum_new_data_count: int = 0,
+    ) -> None:
+        """Initializes an IJEPATrainer object.
+
+        Args:
+            partial_dataloader: A partially instantiated dataloader lacking a provided dataset.
+            partial_optimizer: A partially instantiated optimizer lacking provided parameters.
+            i_jepa_mask_collator: A collator to make masks for I-JEPA models.
+            device: The accelerator device (e.g., CPU, GPU) utilized for training the model.
+            minimum_new_data_count: Minimum number of new data count required to run the training.
+        """
+        super().__init__()
+        self.partial_optimizer = partial_optimizer
+        self.partial_dataloader = partial_dataloader
+        self.i_jepa_mask_collator = i_jepa_mask_collator
+        self.device = device
+        self.logger = logger
+        self.max_epochs = max_epochs
+        self.minimum_dataset_size = minimum_dataset_size
+        self.minimum_new_data_count = minimum_new_data_count
+
+    def on_data_users_dict_attached(self) -> None:
+        self.image_data_user: ThreadSafeDataUser[RandomDataBuffer] = self.get_data_user(BufferNames.IMAGE)
+
+    def on_model_wrappers_dict_attached(self) -> None:
+        self.context_encoder: ModelWrapper[IJEPAEncoder] = self.get_training_model(ModelNames.I_JEPA_CONTEXT_ENCODER)
+        self.predictor: ModelWrapper[IJEPAPredictor] = self.get_training_model(ModelNames.I_JEPA_PREDICTOR)
+        self.target_encoder: ModelWrapper[IJEPAEncoder] = self.get_training_model(ModelNames.I_JEPA_TARGET_ENCODER)
+
+        # Since the model is swapped between the inference and training threads each time it is trained,
+        # the model and optimizer are built within the `train()` method.
+        # The following is the initial state generation of the optimizer.
+        self.optimizer_state = self.partial_optimizer(
+            itertools.chain(self.context_encoder.parameters(), self.predictor.parameters())
+        ).state_dict()
+
```

Comment:
これは実装を確認して欲しいんだけど、TargetEncoderとContextEncoderって同じパラメータからスタートすると思う。ただ、ちょっと AMI Systemの実装的に、instantiateをそれぞれに対して呼び出すことになりそうで、初期パラメータの同一性を保証できない。だからon_model_wrappers_dict_attachedの中でパラメータのコピーをして欲しい。

また、コピーする時には Target EncoderからCortextEncoderにコピーすることを注意してほしい。
 Target Encoderは 推論スレッドと学習スレッドに**2重に** 存在していて、TargetEncoderの方のパラメータを変えちゃうと同期処理を手動で呼ばなければいけなくなってしまうんだ。
```suggestion
        for context_encoder_param, target_encoder_param in zip(self.context_encoder.paramters(), self.target_encoder.parameters()):
            context_encoder_param.copy_(target_encoder_param)
```

Reply:
https://github.com/MLShukai/ami-vconf24/pull/47/commits/9845eabb6a827596aa155f3c822283c719211230 で追加